LSPS3/README.md

```diff
@@ -0,0 +1,317 @@
+# LSPS3 JIT Channel Negotiation
+
+|Name    | `base_api`               |
+|--------|--------------------------|
+|Version | 1                        |
+
+## Motivation
+
+A "JIT Channel" is a 0-conf channel opened in response to an incoming payment
+from the public network to a client, via the LSP.
+This allows a client with no Lightning channels to start receiving on
+Lightning, and have the cost of their inbound liquidity be deducted from their
+first received payment.
+
+### Trust Models
+
+As of this version, the trust model is that the LSP trusts clients to actually
+claim their payment once the channel is opened and the payment is forwarded.
+
+If the client does not claim the payment, and the funding transaction confirms,
+then the LSP will have its funds locked in a channel that was not paid for.
+Even if the LSP disables all other use of the channel until the payment paying
+for the channel is claimed, the client may then refuse all mutual close attempts
+in retaliation (by disconnecting), and force the LSP to unilaterally close,
+which locks the funds until the `to_self_delay` (indicated by the client, and
+imposed on the LSP) has passed.
+
+LSPs that implement this API SHOULD include mitigations to allow them to
+determine if clients can be trusted, or to detect and protect against attacks
+on the LSP.
+How those mitigations are performed is beyond the scope of this specification.
+
+### Amounts
+
+All amounts in this protocol are specified in millisatoshi.
+
+All millisatoshi values MUST be represented as a string and MUST NOT be
+represented as integer values.
+
+> **Rationale** JSON numbers are specified to only fit IEEE double-precision
+> numbers, which have a 53 bit mantissa.
+> The maximum number of millisatoshis (21 million BTC) cannot be specified in
+> 53 bits.
+> In addition, actual JSON parsers may use a smaller size for parsed integers;
+> for instance, as signed 32 bit numbers.
+> The largest safe value is 2,147,483,647; 2,147,483,647msat = BTC0.021,474,836,47
+> which is too low.
+
+### Actors
+
+'LSP' is the API provider.
+'Client' is the API client.
+'Payer' is who pays for the initial receive of the client.
+
+## Flow (LSP Trusts Client Model)
+
+Overview:
+
+* The client determines that it would like to receive over Lightning, but
+  it has no inbound capacity, and the client cannot pre-buy this capacity
+  (for instance, it also has no outbound capacity or onchain funds).
+* The client determines the parameters of a particular LSP via a `GET`
+  request.
+* The client requests for a JIT channel, specifying how much it will
+  receive, via a `POST` request.
+* The LSP returns an SCID, which identifies this request to the LSP.
+* The client generates an invoice, which includes the above SCID and the
+  LSP node ID as a routehint.
+* The client hands the invoice to whoever it will receive funds from.
```

Comment:
This sounds like it assumes the client is generating an invoice with an amount (so that it would know it doesn't have sufficient inbound liquidity and would need to do the above steps to create a special invoice with this JIT LSP flow).

How would 0 amount invoices work where the send selects the amount to be paid?

Reply:
Would it be recommended that a client that generates a 0 amount invoice to check its inbound liquidity and if it (somehow) thinks the amount it might receive will be more than its inbound liquidity, to go ahead and make a LSP JIT request?

Related, I do not see how keysend payments would be supported. And the async payments proposal uses keysend. So, how would async payments work with this proposal?

Reply:
> How would 0 amount invoices work where the send selects the amount to be paid?

What happens if the payer selects an amount that is less than the `opening_fee`?  LSP would not want to open the channel then (not enough to pay for the channel open), and would have to hold the amount (which is not differentiatable from a channel jamming attack) until the client gets another payer (or has the same payer pay further with a different payment) and the sum total exceeds the `opening_fee`.

>Related, I do not see how keysend payments would be supported. And the async payments proposal uses keysend. So, how would async payments work with this proposal?

keysend requires that the receiver be findable on the public network, since there is no invoice issued by the receiver and thus no way to tell the sender what the LSP of the receiver is.

If the client has no channels yet, it cannot possibly be findable on the public network, so JIT channels are not inherently compatible with that use-case: without any channels, the client is not on the public network and keysend to it is impossible.

> And the async payments proposal uses keysend. 

Seriously??? That makes no sense.  Is this the same as the Offline Receive scheme by Matt?  Because that only needs onion messages, not keysend.  That can be made to work with an invoice and no keysend scheme, since the only needed information is some node that is findable on the public network, and the invoice has to have a routehint pointing to the receiver LSP, so that is enough for the payer (or the payer LSP, if the payer is not itself a public node) to be able to send an onion message to the receiver LSP to say "I wanna make an async payment to this receiver with this SCID from you".

Maybe the keysend usage is only the initial implementation, but as I understand it the Matt Corallo Offline Receive scheme does not inherently require keysend.

Reply:
`payment_size + opening_fee`, perhaps?  But that does mean the client has to pre-commit to a specific amount, possibly.

Reply:
I think I want to defer the issue of variable-amount invoice for the first payment, since it does not seem to be something that is easy to consider now, and it may be better to have *some* imperfect, simpler spec sooner than a more complex spec later.

Reply:
> Seriously??? That makes no sense. Is this the same as the Offline Receive scheme by Matt? Because that only needs onion messages, not keysend. 

Yes async payments is the name of Matt's "offline receive." Yes it uses keysend. This is required to do trust-minimized payments (can't allow LSP to issue invoices which could be stolen). If you believe it is possible with invoices, I encourage you to engage with Matt, Val, Jeff, Bastien, Richard Myers, and others working on async payments to discuss.

Assuming keysend is the case, however, the sender WILL know the receiver's LSP as part of the async payments protocol. So hopefully async payments can be accommodated with 0conf JIT, otherwise this is at best a temporary solution, and we'd need something different once async payments are being deployed.

In any case, it seems imperative to me that the JIT 0conf protocol will work with the async payments protocol, as the latter would be expected to become the predominant way for mobile and web wallets to receive payments (as they don't have reliable internet connectivity).

Reply:
Then there should not be anything blocking that definition; the specs as laid out here does not care how the client learns the preimage (whether it already knows it, or is given it in an encrypted onion only it can decode).